### PR TITLE
7 hnz clientserver reconnection issues

### LINKED ISSUE #7: HNZ client/server reconnection issues
**Describe the bug**
During the resolution of https://github.com/fledge-power/fledge-south-hnz/issues/60, several issues were found when trying to disconnect and reconnect a TCP socket using `HNZClient`/`HNZServer`.

This ticket tracks the resolution of the following issues:

- Crash when trying to delete a `HNZClient`/`HNZServer` object that was not stopped (`m_ThreadAutomate` still running).
- Crash in `HNZClien::stop()`/ `HNZServer::stop()` when `m_ThreadAutomate` was never started.
- Inability to start a `HNZClient`/`HNZServer` again after stopping it.
- Stopping a server connection in `TcpConnection` leaves the listening server socket open forever.
- Memory leak when deleting a `HNZClient`/`HNZServer` object.
- Connection stuck never sending any message due to uninitialized variable in `VoieHNZ`.
- Connection lost at TCP level not detected correctly.
- Unhandled case where socket `send()` would not be able to send the whole message in one call.
- Sockets representing connection with the client never being closed until end of application lifetime.

### PR DESCRIPTION
**Pull request by FlorentP42**
Closes #7

### CODE REVIEW

File: src/hnz/TcpConnexion.cpp [hunk 1]
```diff
@@ -78,26 +78,53 @@ void TcpConnexion::iTCPConnecteServeur(int port) {
...
-  if(sent == -1){
-    // Display error if any
-    perror("send:");
-    perror(strerror(errno));
-  }
+  int iNbSent = 0;
+  int iNbTotalSent = 0;
+  // Message send loop
+  do {
+    // printf("TcpConnexion::vSend - socket %d - Try sending data...\n", socketfd);
```
Comment (YmaIneo) [inline src/hnz/TcpConnexion.cpp:110]: Is this comment useful?

File: src/hnz/TcpConnexion.cpp [hunk 2]
```diff
@@ -106,18 +133,36 @@ bool TcpConnexion::vAssembleTrame() {
...
 
   bool trameRecue = false;
-  // Boucle de lecture du message
+  // Message reception loop
   do {
     // Lecture d'un octet sur le socket
     iNbLus = recv(socketfd, &c, 1, 0);
     // Test retour lecture socket
     if (iNbLus > 0) {
+      // printf("TcpConnexion::vAssembleTrame - socket %d - Received byte: %02X\n", socketfd, static_cast<unsigned int>(c) & 0xFF);
```
Comment (YmaIneo) [inline src/hnz/TcpConnexion.cpp:142]: Is this comment useful?

File: src/hnz/TcpConnexion.cpp [hunk 3]
```diff
@@ -106,18 +133,36 @@ bool TcpConnexion::vAssembleTrame() {
...
       if (m_pVoie->bAjouteCar(c)) {
         trameRecue = m_pVoie->vTraiteEvt(EVT_VOIE_TRAME_HNZ);
+        // if (trameRecue) {
+        //   printf("TcpConnexion::vAssembleTrame - socket %d - Received full frame\n", socketfd);
+        // }
```
Comment (YmaIneo) [inline src/hnz/TcpConnexion.cpp:147]: Are these comments useful?

File: src/hnz/TcpConnexion.cpp [hunk 4]
```diff
@@ -127,18 +172,54 @@ bool TcpConnexion::vAssembleTrame() {
...
   loop = false;
-  return shutdown(socketfd, SHUT_RDWR);
+  if (socketfd != -1) {
+    int res = shutdown(socketfd, SHUT_RDWR);
+    if (res == -1) {
+      printf("TcpConnexion::stop - socket %d - Error in shutdown: %s\n", socketfd, strerror(errno));
+    }
+    // else {
+    //   printf("TcpConnexion::stop - socket %d - Shutdown socket\n", socketfd);
+    // }
```
Comment (YmaIneo) [inline src/hnz/TcpConnexion.cpp:203]: Are these comments useful?

File: src/hnz/TcpConnexion.cpp [hunk 5]
```diff
@@ -127,18 +172,54 @@ bool TcpConnexion::vAssembleTrame() {
...
+    res = close(socketfd);
+    if (res == -1) {
+      printf("TcpConnexion::stop - socket %d - Error in close: %s\n", socketfd, strerror(errno));
+    }
+    // else {
+    //   printf("TcpConnexion::stop - socket %d - Closed socket\n", socketfd);
+    // }
```
Comment (YmaIneo) [inline src/hnz/TcpConnexion.cpp:210]: Are these comments useful?

File: src/hnz/TcpConnexion.cpp [hunk 6]
```diff
@@ -127,18 +172,54 @@ bool TcpConnexion::vAssembleTrame() {
...
+    socketfd = -1;
+  }
+  if (server_fd != -1) {
+    int res = close(server_fd);
+    if (res == -1) {
+      printf("TcpConnexion::stop - socket %d - Error in close listen: %s\n", server_fd, strerror(errno));
+    }
+    // else {
+    //   printf("TcpConnexion::stop - socket %d - Closed listen socket\n", server_fd);
+    // }
```
Comment (YmaIneo) [inline src/hnz/TcpConnexion.cpp:220]: Are these comments useful?

File: src/hnz/VoieHNZ.cpp [hunk 7]
```diff
@@ -47,7 +50,12 @@ void VoieHNZ::vInitEnteteTrameData(MSG_TRAME *pTrame, bool bMaj) {
 ****************************************************************************************/
 void VoieHNZ::vEnvoiTrameVersHnz(MSG_TRAME *pTrame) {
   if (!pTrame) return;
+  // printf("VoieHNZ::vEnvoiTrameVersHnz - socket %d - Frame added to stack [%s]\n", connslave->socketfd, frameToStr(pTrame).c_str());
```
Comment (YmaIneo) [inline src/hnz/VoieHNZ.cpp:53]: Is this comment useful?

File: src/hnz/VoieHNZ.cpp [hunk 8]
```diff
@@ -59,6 +67,7 @@ void VoieHNZ::vEnvoiTrameVersHnz() {
     m_bEmissionRecente = false;
     return;
   }
+  // printf("VoieHNZ::vEnvoiTrameVersHnz - socket %d - Frame taken from stack [%s]\n", connslave->socketfd, frameToStr(pTrame).c_str());
```
Comment (YmaIneo) [inline src/hnz/VoieHNZ.cpp:70]: Is this comment useful?

File: src/hnz/VoieHNZ.cpp [hunk 9]
```diff
@@ -129,11 +138,14 @@ bool VoieHNZ::bAjouteCar(unsigned char ub) {
...
 
 void VoieHNZ::vGereAutomate() {
+  printf("VoieHNZ::vGereAutomate - socket %d - Timers management thread started\n", connslave->socketfd);
+
   while (!stop_flag) {
     void *pv;
     int iEvtTimer;
     bool timer = m_AutomateHNZ->bGetNextTimer(iEvtTimer, pv);
     if (timer) {
+      // printf("VoieHNZ::vGereAutomate - socket %d - Processing event of type %d\n", connslave->socketfd, iEvtTimer);
```
Comment (YmaIneo) [inline src/hnz/VoieHNZ.cpp:148]: Is this comment useful?

File: src/hnz/VoieHNZ.cpp [hunk 10]
```diff
@@ -145,11 +157,15 @@ void VoieHNZ::vGereAutomate() {
     // Get the time until next timer expiration
     uint delay = m_AutomateHNZ->uiGetTimeOut();
     if (delay == 0x7FFFFFF) {
+      //printf("VoieHNZ::vGereAutomate - socket %d - No timer in list\n", connslave->socketfd);
```
Comment (YmaIneo) [inline src/hnz/VoieHNZ.cpp:160]: Is this comment useful?

File: src/hnz/VoieHNZ.cpp [hunk 11]
```diff
@@ -145,11 +157,15 @@ void VoieHNZ::vGereAutomate() {
...
       // Use 50ms as default delay as this is the minimum time ever set for a new timer
       delay = DELAI_INTER_TRAME_MS;
     }
+    //printf("VoieHNZ::vGereAutomate - socket %d - Waiting for %d ms\n", connslave->socketfd, delay);
```
Comment (YmaIneo) [inline src/hnz/VoieHNZ.cpp:164]: Is this comment useful?

File: .gitignore [hunk 12]
```diff
@@ -50,3 +50,4 @@ modules.order
 Module.symvers
 Mkfile.old
 dkms.conf
+.vscode/
```
[the remark this answers is not in the document]
Comment (FlorentP42) [reply .gitignore:53]: This ignores vscode config folder so it is not accidentally commited into the repository

**Comment by FlorentP42** [comment]
All commented prints may be uncommented for advanced debugging in case random connection issues are encountered again in the future.

**Pull request merged**